spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DeadLetterPublishingRecovererFactory.java

```diff
@@ -253,13 +268,17 @@ private static String getRecordInfo(ConsumerRecord<?, ?> cr) {
 	 * and if it doesn't it sets -1, to allow the Producer itself to assign a partition to the record.</p>
 	 *
 	 * <p>Subclasses can inherit from this method to override the implementation, if necessary.</p>
+	 * The destination partition can also be customized using {@link #setPartitionResolver(BiFunction)}.
 	 *
 	 * @param cr The original {@link ConsumerRecord}, which is to be forwarded to DLT
 	 * @param nextDestination The next {@link DestinationTopic}, where the consumerRecord is to be forwarded
-	 * @return An instance of {@link TopicPartition}, specifying the topic and partition, where the cr is to be sent
+	 * @return An instance of {@link TopicPartition}, specifying the topic and partition, where the cr is to be sent.
+	 * @see #setPartitionResolver(BiFunction)
 	 */
 	protected TopicPartition resolveTopicPartition(final ConsumerRecord<?, ?> cr, final DestinationTopic nextDestination) {
-		return new TopicPartition(nextDestination.getDestinationName(), cr.partition());
+		String nextTopic = nextDestination.getDestinationName();
+		Integer partition = this.partitionResolver.apply(cr, nextTopic);
+		return new TopicPartition(nextTopic, partition == null ? -1 : partition);
```

Comment:
Just a question for clarification: how is it going to behave if `has fewer partitions than the main topic`?
I believe in most cases people are going to rely on a default resolver, which may lead to non-existing partition.

Thanks